src/ide/code_lens/tests.rs

```diff
@@ -197,69 +191,119 @@ impl TestRunner {
-                                },
-                                true,
-                                file_state,
-                            );
-                        };
-                    }
+    collect_functions(db, module, make_function_code_lens(db, origin_file, &file_url, file_state));
+
+    let Ok(modules) = db.module_submodules_ids(module) else { return };
+
+    for submodule in modules.iter().copied() {
+        let has_tests = if db.is_submodule_inline(submodule) {
+            let tests_count = file_state.len();
+
+            collect_tests(
+                db,
+                ModuleId::Submodule(submodule),
+                origin_file,
+                file_url.clone(),
+                file_state,
+            );
+
+            // Append mod only if it contains tests.
+            tests_count != file_state.len()
+        } else {
+            has_any_test(db, ModuleId::Submodule(submodule))
+        };
+
+        if !has_tests {
+            continue;
+        }
+
+        let ptr = submodule.stable_ptr(db).untyped();
+
+        let Some(position) = get_position(db, origin_file, ptr) else {
+            continue;
+        };
+
+        if ptr.file_id(db) != origin_file {
+            continue;
+        }
+
+        make_code_lens(
+            TextDocumentPositionParams {
+                position,
+                text_document: TextDocumentIdentifier { uri: file_url.clone() },
+            },
+            true,
+            file_state,
+        );
+    }
+}
+
+fn has_any_test(db: &AnalysisDatabase, module: ModuleId) -> bool {
+    if collect_functions(db, module, |_| ControlFlow::Break(())).is_break() {
+        return true;
+    }
+
+    let Ok(modules) = db.module_submodules_ids(module) else { return false };
+
+    modules.iter().copied().map(ModuleId::Submodule).any(|submodule| {
+        collect_functions(db, submodule, |_| ControlFlow::Break(())).is_break()
+            || has_any_test(db, submodule)
+    })
+}
+
+fn get_position(db: &AnalysisDatabase, file: FileId, ptr: SyntaxStablePtrId) -> Option<Position> {
+    ptr.lookup(db)
+        .span_start_without_trivia(db)
+        .position_in_file(db, file)
+        .map(|position| position.to_lsp())
+}
+
+fn collect_functions(
+    db: &AnalysisDatabase,
+    module: ModuleId,
+    mut on_test: impl FnMut(SyntaxStablePtrId) -> ControlFlow<(), ()>,
+) -> ControlFlow<(), ()> {
+    if let Ok(functions) = db.module_free_functions(module) {
+        for function in functions.values() {
+            if let Some(test_attr) = function.find_attr(db, "test") {
+                match on_test(test_attr.stable_ptr().untyped()) {
+                    b @ ControlFlow::Break(_) => return b,
+                    ControlFlow::Continue(_) => continue,
                 }
             }
-            _ => {}
         }
     }
+
+    ControlFlow::Continue(())
+}
+
+fn make_function_code_lens(
```

Comment:
```suggestion
fn code_lens_factory(
```

or something other that suggests what it actually is